Added recognition of EOF at LL(1) parsers.

kuniss opened the pull request:
Made the old behavior for Oberon2 which allows arbitrary trailing content before EOF optional and the new behavior the default.
The option of allowing arbitrary content can be enabled by passing the command line option '-t' to the generated compiler.
This solves GH issue #11.

kuniss committed:
Added recognition of EOF at LL(1) parsers.

Made the old behavior for Oberon2 which allows arbitrary trailing
 content before EOF optional and
 the new behavior the default.
 The option of allowing arbitrary content can be enabled by passing
 the command line option '-t' to the generated compuiler.
 Added according tests at test/ell1.d.
 This solves GH issue #11.

kuniss committed:
Slight formating correction to satisfy style check

linkrope commented on fix/epsilon/ell1gen.fix.d:
```diff
@@ -253,10 +256,27 @@ void Compile(Input input, bool info_, bool verbose, bool write)
+    {
+        if (Tok == sepTok) 
+            continue;
+        else {
```
Use Allman style!

BTW: avoid the strange loop:
````
while (Tok == sepTok)
    S.Get(Tok);
if (Tok != endTok)
{
    ++ErrorCounter;
    error!"syntax error, unexpected content before end of file\n%s"(S.Pos);
}
````

kuniss replied:
That's from my Java background... I will do.

linkrope commented on fix/epsilon/ell1gen.fix.d:
```diff
@@ -207,6 +208,8 @@ void main(string[] args)
                 "info|i", "Show heap usage information.", &info,
                 "verbose|v", "Print verbose parser error messages.", &verbose,
                 "write|w", "Toggle default for writing output.", &write,
+                "trailingTokens|t", "allowing arbitrary trailing content before EOF (Oberon2 like behavior)", 
```
Use consistent term! It's
- ignore content before EOF
- trailing tokens
- training content before EOF

I prefer "content", because both approaches don't use the lexer for the trailing stuff.

(BTW: it's not just Oberon 2, but already Oberon)

kuniss replied:
Not relevant anymore, we go for a backward incompatible change. See below.

linkrope commented on fix/epsilon/ell1gen.fix.d:
```diff
@@ -207,6 +208,8 @@ void main(string[] args)
                 "info|i", "Show heap usage information.", &info,
                 "verbose|v", "Print verbose parser error messages.", &verbose,
                 "write|w", "Toggle default for writing output.", &write,
+                "trailingTokens|t", "allowing arbitrary trailing content before EOF (Oberon2 like behavior)", 
```
This is an elegant way of hiding the LL-specific option from the rest of the compiler generator.
However, the option is more like "generate a compiler with an Oberon-style parser" than "ignore trailing content for this specific input".

I don't see, how this option can be made a compiler-generator option in an elegant way. So, I would recommend no option for this behavior at all. Just remove the condition from `recognizeEof`.

kuniss replied:
@linkrope, just to have no misunderstanding between us: 
The option was added to the generated compiler, not the generator! 
The first time, I read your note I thought, you are mixing up generated compiler and generator. Now, I'm not sure whether I maybe get our note wrongly...

kuniss replied:
Not relevant anymore, we go for a backward incompatible change. See below.

linkrope commented on fix/epsilon/ell1gen.fix.d:
```diff
@@ -253,10 +256,27 @@ void Compile(Input input, bool info_, bool verbose, bool write)
         S.Get(Tok);
         $(V1);
+        if (!ignoreContentBeforeEof)
+            recognizeEof();
```
Omit parentheses:
```
recognizeEof;
```

kuniss replied:
Is this D style? I will do, of course.

linkrope commented on test/ell1.d:
```diff
@@ -11,3 +12,37 @@ unittest
+        run!"cd %s && echo x yy zzz | ./NoEOF -t"(directory)
+            .shouldPassWith("Done");
+    }
+}
```
Even GitHub complains about the missing newline...

kuniss replied:
"He did it again..." :smile:

linkrope commented on test/ell1.d:
```diff
@@ -11,3 +12,37 @@ unittest
+    with (sandbox)
+    {
+        const eag = `
+            NoEOF<+ 'Done': Done>: 'x'.
```
While the grammar is nicely reduced, a grammar for `a^n b^n`  would better motivate the issue:
`a b b` is a valid input.

kuniss replied:
Indeed, it's a more pathological case at all.

kuniss replied:
The grammar example `a^n b^n` does not work for reproducing the issue behavior. To trigger the behavior, the last symbol in the grammar must be mandatory and not part of a repetition or option. 
At least, I could not reproduce this behavior with the mentioned grammar. So, I stay with the original minimal grammar.

linkrope replied:
I mean the classical context-free grammar
```
S: | 'a' S 'b'.
```

kuniss replied:
Got it.

linkrope commented on test/ell1.d:
```diff
@@ -11,3 +12,37 @@ unittest
+
+        run!"cat <<EOF | ./gamma --output-directory %s%sEOF"(directory, eag)
+            .shouldPassWith("NoEOF grammar is SLAG");
+        run!"cd %s && echo x yy zzz | ./NoEOF -t"(directory)
```
`-t` is for manual input; a script should use the long option `--tralingTokens`

kuniss replied:
Yes, better readability in the tests.

kuniss replied:
Not relevant anymore, we go for a backward incompatible change. See below.
This test will disappear either.

linkrope reviewed (changes requested):
I propose a breaking change without any option for the olb behavior.

kuniss commented:
> I propose a breaking change without any option for the old behavior.

So, is your proposal to remove the old behavior at all and put the trailing content at the Oberon examples in single line comments?

linkrope commented:
> > I propose a breaking change without any option for the old behavior.
> 
> So, is your proposal to remove the old behavior at all and put the trailing content at the Oberon examples in single line comments?

Yes! Because the elegant solution is to add a compiler option, while it should have been a genrator option.
I want to generate an Oberon compiler, but then the option is only valid, if I generate an LL parser. A dependent option.
Placing the option into the LL parser code avoids this problem, but then you have to select this option for every run of the generated Oberon compiler.

kuniss commented:
> > > I propose a breaking change without any option for the old behavior.
> > 
> > 
> > So, is your proposal to remove the old behavior at all and put the trailing content at the Oberon examples in single line comments?
> 
> Yes! Because the elegant solution is to add a compiler option, while it should have been a genrator option. I want to generate an Oberon compiler, but then the option is only valid, if I generate an LL parser. A dependent option. Placing the option into the LL parser code avoids this problem, but then you have to select this option for every run of the generated Oberon compiler.

Agree!

kuniss committed:
Made EOF recognition mandatory.

By that removed previously introduced  "-t" option again.
Adjusted tests for that.

kuniss committed:
Put trailing Oberon System commands in comments

as no trailing content is allowed anymore before EOF,  except comments.

kuniss committed:
Improved example for GH #11 test

to make it more obvious, as suggested by @linkrope.

linkrope reviewed (approved):
The token name "end" tries to abstract from "end of file", "end of text"...
Here, we should also consistently say "end" instead of "EOF".

kuniss committed:
Improved naming and errior message

as suggested by @linkrope.

kuniss commented:
> The token name "end" tries to abstract from "end of file", "end of text"... Here, we should also consistently say "end" instead of "EOF".

I improved the naming and the error message. :smile: 
But I found it better to reference both "end of file" and "end of text" in the error message with a preference to "end of file" as this will normally happen, right?

linkrope commented:
> But I found it better to reference both "end of file" and "end of text" in the error message with a preference to "end of file" as this will normally happen, right?

OK. I would just have said "unexpected content at the end".
